Core/Modes/Matchmode.cs

```diff
@@ -1,43 +1,132 @@
             _logger.LogInformation("Matchmode starting");
             _modeBase.Start();
+            Task overlayWebsocketTask = _broadcastServer.Listen();
             while (!_stopToken.ShouldStop)
             {
-                // TODO match main loop goes here
-                await Task.Delay(TimeSpan.FromMilliseconds(100));
+                _loopCancelToken = new CancellationTokenSource();
+                try
+                {
+                    await Loop(_loopCancelToken.Token);
+                }
+                catch (OperationCanceledException)
+                {
+                    _logger.LogInformation("Match loop cancelled");
+                }
             }
+            await _broadcastServer.Stop();
+            await overlayWebsocketTask;
             _logger.LogInformation("Matchmode ended");
         }
 
+        private async Task Loop(CancellationToken cancellationToken)
+        {
+            var teams = new Teams
+            {
+                Blue = ImmutableList.Create(MatchTesting.TestVenonatForOverlay),
+                Red = ImmutableList.Create(MatchTesting.TestVenonatForOverlay),
+            };
+            await Task.Delay(TimeSpan.FromSeconds(3), cancellationToken);
+
+            IMatchCycle match = new CoinflipMatchCycle(_loggerFactory.CreateLogger<CoinflipMatchCycle>());
+            Task setupTask = match.SetUp(new MatchInfo(teams.Blue, teams.Red), cancellationToken);
+            await _overlayConnection.Send(new MatchCreatedEvent(), cancellationToken);
+            await _overlayConnection.Send(new MatchBettingEvent(), cancellationToken);
+            await _overlayConnection.Send(new MatchModesChosenEvent(), cancellationToken); // TODO
+            await _overlayConnection.Send(new MatchSettingUpEvent
+            {
+                MatchId = 1234,
+                Teams = teams,
+                BettingDuration = _matchmodeConfig.DefaultBettingDuration.TotalSeconds,
+                RevealDuration = 0,
+                Gimmick = "speed",
+                Switching = "never",
+                BattleStyle = "singles",
+                InputOptions = new InputOptions
+                {
+                    Moves = new MovesInputOptions
+                    {
+                        Policy = "always",
+                        Permitted = ImmutableList.Create("a", "b", "c", "d")
+                    },
+                    Switches = new SwitchesInputOptions
+                    {
+                        Policy = "never",
+                        Permitted = ImmutableList<string>.Empty,
+                        RandomChance = 0
+                    },
+                    Targets = new TargetsInputOptions
+                    {
+                        Policy = "disabled",
+                        Permitted = ImmutableList<string>.Empty,
+                        AllyHitChance = 0
+                    },
+                },
+                BetBonus = 35,
+                BetBonusType = "bet",
```

Comment:
enum

Reply:
I suggest ignoring the bet bonus stuff for now, because I'm not sure I want to keep bet bonuses the way they currently are anyway. I think it's better if the two kinds of bet bonuses were distinct fields like `bet_bonus` and `profit_bonus` instead of one field whose meaning gets interpreted based on the contents of another field. And that other field doesn't even get persisted right now :(